pkg/task/registration.go

```diff
@@ -0,0 +1,165 @@
+package task
+
+import (
+	"sync"
+
+	"strings"
+
+	"github.com/Peripli/service-broker-proxy/pkg/platform"
+	"github.com/Peripli/service-broker-proxy/pkg/sm"
+	"github.com/sirupsen/logrus"
+)
+
+const ProxyBrokerPrefix = "sm-proxy-"
+
+type SBProxyRegistration struct {
+	group          *sync.WaitGroup
+	platformClient platform.Client
+	smClient       sm.Client
+	proxyHost      string
+}
+
+type serviceBrokerReg struct {
+	platform.ServiceBroker
+	SmID string
+}
+
+func New(group *sync.WaitGroup, platformClient platform.Client, smClient sm.Client, proxyHost string) *SBProxyRegistration {
+	return &SBProxyRegistration{
+		group:          group,
+		platformClient: platformClient,
+		smClient:       smClient,
+		proxyHost:      proxyHost,
+	}
+}
+
+func (r SBProxyRegistration) Run() {
+	r.group.Add(1)
+	defer r.group.Done()
+	r.run()
+}
+
+func (r SBProxyRegistration) run() {
+	logrus.Debug("Running broker registration task...")
+	registeredBrokers, err := r.platformClient.GetBrokers()
+	if err != nil {
+		logrus.WithError(err).Error("An error occurred while obtaining already registered brokers")
+		return
+	}
+
+	brokersFromPlatform := make([]serviceBrokerReg, 0, len(registeredBrokers))
+	for _, broker := range registeredBrokers {
+		if !r.isBrokerProxy(broker) {
+			logrus.WithFields(logFields(&broker)).Debug("Registration task SKIPPING registered broker as is not recognized to be proxy broker...")
+			continue
+		}
+
+		if f, isFetcher := r.platformClient.(platform.Fetcher); isFetcher {
+			if err := f.Fetch(&broker); err != nil {
```

Comment:
may fetch a broker that should be removed
generally this function should do this:
1. get the brokers from the platform
2. get the broker from SM
3. unregister those that are only in the platform
4. register those that are only in SM
5. update the catalog of the rest

it would be nice if the structure of this function reflects this flow, i.e. exact in a separate function the implementation of each step
Also the term "Fetch" is too generic, consider renaming to something more specific, e.g. FetchCatalog or UpdateCatalog

Reply:
we will do one more loop though :/